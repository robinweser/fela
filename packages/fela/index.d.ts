declare module "fela" {
  import * as CSS from 'csstype';
  import { TRuleType, TKeyframeType, TFontType, TStaticType, TClearType } from 'fela-utils';

  export type TRuleProps = {};
  export type TRule<T = TRuleProps> = (props: T, renderer: IRenderer) => IStyle
  export type TKeyFrame<T = TRuleProps> = (props: T, renderer: IRenderer) => {
    from?: IStyle,
    to?: IStyle,

    [persent: string]: IStyle | undefined;
  };

  type TRendererCreator = (config?: IConfig) => IRenderer;
  type TPlugin = (style: IStyle) => IStyle; //http://fela.js.org/docs/advanced/Plugins.html
  type TEnhancer = (renderer: IRenderer) => IRenderer; //http://fela.js.org/docs/advanced/Enhancers.html

  type TSubscribeMessageType = TRuleType | TKeyframeType | TFontType | TStaticType | TClearType

  interface ISubscribeMessage {
    type: TSubscribeMessageType;
  }
  interface ISubscribeRuleOrStaticObjectMessage extends ISubscribeMessage { static?: boolean; declaration: string; selector: string; media: string; }
  interface ISubscribeKeyframesMessage extends ISubscribeMessage { name: string; keyframe: string; }
  interface ISubscribeFontFaceMessage extends ISubscribeMessage { fontFamily: string; fontFace: string; }
  interface ISubscribeStaticStringMessage extends ISubscribeMessage { css: string; }
  interface ISubscribeClearMessage extends ISubscribeMessage { }

  interface IRenderer {
    renderRule<T = TRuleProps>(rule: TRule<T>, props: T): string
    renderKeyframe<T = TRuleProps>(keyFrame: TKeyFrame<T>, props: T): string;
    renderFont<T = TRuleProps>(family: string, files: Array<string>, props: T): void;
    renderStatic(style: string, selector?: string): void;
    renderStatic(style: IStyle, selector: string): void;
    renderToString(): string;
    subscribe(event: (msg: ISubscribeRuleOrStaticObjectMessage | ISubscribeKeyframesMessage | ISubscribeFontFaceMessage | ISubscribeStaticStringMessage | ISubscribeClearMessage) => void): { unsubscribe: () => void; }
    clear(): void;
  }

  //http://fela.js.org/docs/advanced/RendererConfiguration.html
  interface IConfig {
    plugins?: Array<TPlugin>;
    keyframePrefixes?: Array<string>;
    enhancers?: Array<TEnhancer>;
    mediaQueryOrder?: Array<string>;
    selectorPrefix?: string;
    specificityPrefix?: string;
    filterClassName?: (className: string) => boolean;
    devMode?: boolean;
  }

  export interface IStyle extends CSS.Properties<string | number> {
    // for selectors and pseudo classes use fela-plugin-typescript
  }

  function createRenderer(config?: IConfig): IRenderer;

  function combineRules<A, B>(a: TRule<A>, b: TRule<B>): TRule<A & B>
  function combineRules<A, B, C>(
    a: TRule<A>,
    b: TRule<B>,
    c: TRule<C>,
  ): TRule<A & B & C>
  function combineRules(...rules: Array<TRule>): TRule

  function enhance(...enhancers: Array<TEnhancer>): (rendererCreator: TRendererCreator) => TRendererCreator;
}

declare module "fela-dom" {
  import { IRenderer } from 'fela';

  function render(renderer: IRenderer): void;
  function rehydrate(renderer: IRenderer): void;
  function renderToMarkup(renderer: IRenderer): string;
  function renderToSheetList(renderer: IRenderer): {
    css: string,
    type: 'RULE' | 'KEYFRAME' | 'FONT' | 'STATIC',
    media?: string,
    support?: boolean,
  }[];
}

declare module "fela-tools" {
  import { TRule, TRuleProps, IStyle, IRenderer } from "fela";

  export type TMultiRuleObject<Props = TRuleProps, Styles = {}> = {[key in keyof Styles]: TRule<Props> | IStyle}
  export type TMultiRuleFunction<Props = TRuleProps, Styles = {}> = (props: Props, renderer: IRenderer) => TMultiRuleObject<Props, Styles>
  export type TMultiRule<Props = TRuleProps, Styles = {}> = TMultiRuleObject<Props, Styles> | TMultiRuleFunction<Props, Styles>

  export type TPartialMultiRuleObject<Props = TRuleProps, Styles = {}> = Partial<TMultiRuleObject<Props, Styles>>
  export type TPartialMultiRuleFunction<Props = TRuleProps, Styles = {}> = (props: Props, renderer: IRenderer) => TPartialMultiRuleObject<Props, Styles>
  export type TPartialMultiRule<Props = TRuleProps, Styles = {}> = TPartialMultiRuleObject<Props, Styles> | TPartialMultiRuleFunction<Props, Styles>

  export type TNormalizedMultiRule<Props = TRuleProps, Styles = {}> = (props: Props, renderer: IRenderer) => {[key in keyof Styles]: TRule<Props>}

  function combineMultiRules<A, SA, B, SB>(
    a: TMultiRule<A, SA>,
    b: TMultiRule<B, SB>
  ): TNormalizedMultiRule<A & B, SA & SB>
  function combineMultiRules<A, SA, B, SB, C, SC>(
    a: TMultiRule<A, SA>,
    b: TMultiRule<B, SB>,
    c: TMultiRule<C, SC>,
  ): TNormalizedMultiRule<A & B & C, SA & SB & SC>
  function combineMultiRules(...rules: Array<TMultiRule>): TNormalizedMultiRule

  function mapValueToMediaQuery(
    queryValueMap: { [key: string]: string },
    mapper: ((value: string) => object) | string
  ): object;

  function renderToElement(
    renderer: IRenderer,
    mountNode: { textContent: string },
  ): (() => void);

  function renderToString(
    renderer: IRenderer,
  ): string;
}

/**
 * ENHANCERS
 */
declare module "fela-beautifier" {
  import { TEnhancer } from "fela";

  export default function(options?: object): TEnhancer;
}

declare module "fela-identifier" {
  import { TRule, TEnhancer } from "fela";

  interface Configs {
    prefix?: string;
    generator?: (name?: string, index?: number) => string;
  }

  type Identifier = (name?: string) => TRule & {
    className: string;
    toString(): string;
  };

  export default function(configs?: Configs): TEnhancer & Identifier;
}

declare module "fela-layout-debugger" {
  import { TEnhancer } from "fela";

  interface DebuggerOptions {
    mode?: "outline" | "backgroundColor";
    thickness?: number;
  }

  export default function(options?: DebuggerOptions): TEnhancer;
}

declare module "fela-logger" {
  import { TEnhancer } from "fela";

  interface LoggerOptions {
    logCSS?: boolean;
    formatCSS?: boolean;
  }

  export default function(options?: LoggerOptions): TEnhancer;
}

declare module "fela-monolithic" {
  import { TEnhancer } from "fela";

  interface MonolithicOptions {
    prettySelectors?: boolean;
  }

  export default function(options?: MonolithicOptions): TEnhancer;
}

declare module "fela-perf" {
  import { TEnhancer } from "fela";

  export default function(): TEnhancer;
}

declare module "fela-statistics" {
  import { TEnhancer } from "fela";

  export default function(): TEnhancer;
}

declare module "fela-utils" {
  export type TRuleType = "RULE"
  export type TKeyframeType = "KEYFRAME"
  export type TFontType = "FONT"
  export type TStaticType = "STATIC"
  export type TClearType = "CLEAR"

  export const RULE_TYPE: TRuleType
  export const KEYFRAME_TYPE: TKeyframeType
  export const FONT_TYPE: TFontType
  export const STATIC_TYPE: TStaticType
  export const CLEAR_TYPE: TClearType
}

declare module "fela-sort-media-query-mobile-first" {
  import { TEnhancer } from "fela";

  export default function(): TEnhancer;
}

/**
 * PLUGINS
 */
declare module "fela-plugin-bidi" {
  import { TPlugin } from "fela";

  export default function(flowDirection: 'ltr' | 'rtl'): TPlugin;
}

declare module "fela-plugin-custom-property" {
  import { TPlugin } from "fela";

  interface CustomProperties {
    [property: string]: (value: any) => any,
  }

  export default function(properties: CustomProperties): TPlugin;
}

declare module "fela-plugin-embedded" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-extend" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-fallback-value" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-friendly-pseudo-class" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-important" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-isolation" {
  import { TPlugin } from "fela";

  interface IsolationOptions {
    exclude?: string[];
  }

  export default function(options?: IsolationOptions): TPlugin;
}

declare module "fela-plugin-logger" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-named-keys" {
  import { TPlugin } from "fela";

  interface MediaQueryMap {
    [key: string]: string;
  }

  export default function(mediaQueryMap: MediaQueryMap): TPlugin;
}

declare module "fela-plugin-native-media-query" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-placeholder-prefixer" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-fullscreen-prefixer" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-pseudo-prefixer" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-theme-value" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-prefixer" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-rtl" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-simulate" {
  import { TPlugin } from "fela";

  export default function(): TPlugin;
}

declare module "fela-plugin-unit" {
  import { TPlugin } from "fela";

  export type Unit = "ch" | "em" | "ex" | "rem" | "vh" | "vw" | "vmin" | "vmax" | "px" | "cm" | "mm" | "in" | "pc" | "pt" | "mozmm";

  export interface UnitPerProperty {
    [key: string]: string;
  }

  export default function(
    unit?: Unit,
    unitPerProperty?: UnitPerProperty,
    isUnitlessProperty?: (property: string) => boolean,
  ): TPlugin;
}

declare module "fela-plugin-validator" {
  import { TPlugin } from "fela";

  interface Configs {
    logInvalid?: boolean;
    deleteInvalid?: boolean;
    useCSSLint?: boolean | object;
  }

  export default function(configs?: Configs): TPlugin;
}

/**
 * PRESETS
 */
declare module "fela-preset-web" {
  import { TPlugin } from "fela";
  import { Unit, UnitPerProperty } from "fela-plugin-unit";

  type UnitConfig1 = [Unit]
  type UnitConfig2 = [Unit, UnitPerProperty]
  type UnitConfig3 = [Unit, UnitPerProperty, (property: string) => boolean]

  type UnitConfig = UnitConfig1 | UnitConfig2 | UnitConfig3

  export function createWebPreset({ unit }: { unit?: UnitConfig }): TPlugin[];

  const presets: TPlugin[];
  export default presets;
}

declare module "fela-preset-dev" {
  import { TPlugin } from "fela";

  const presets: TPlugin[];
  export default presets;
}
