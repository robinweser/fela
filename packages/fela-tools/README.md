# fela-tools

A set of useful tools to work with Fela.

<img alt="npm version" src="https://badge.fury.io/js/fela-tools.svg"> <img alt="npm downloads" src="https://img.shields.io/npm/dm/fela-tools.svg"> <a href="https://bundlephobia.com/result?p=fela-tools@latest"><img alt="Bundlephobia" src="https://img.shields.io/bundlephobia/minzip/fela-tools.svg"></a>

## Installation
```sh
yarn add fela-tools
```
You may alternatively use `npm i --save fela-tools`.


## Usage
The API is exported by the `fela-tools` package by default e.g.

```javascript
import { StyleSheet } from 'fela-tools'
```

## API Reference

* [`mapValueToMediaQuery(queryValueMap, mapper)`](docs/mapValueToMediaQuery.md)
* [`renderToString(renderer)`](docs/renderToString.md)
* [StyleSheet](docs/StyleSheet.md)

## License
Fela is licensed under the [MIT License](http://opensource.org/licenses/MIT).<br>
Documentation is licensed under [Creative Commons License](http://creativecommons.org/licenses/by/4.0/).<br>
Created with â™¥ by [@robinweser](http://weser.io) and all the great contributors.
