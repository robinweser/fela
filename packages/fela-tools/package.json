{
  "name": "fela-tools",
  "version": "11.4.0",
  "description": "Useful tools for working with Fela",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "sideEffects": false,
  "files": [
    "LICENSE",
    "README.md",
    "lib/**",
    "es/**"
  ],
  "repository": "https://github.com/robinweser/fela/",
  "keywords": [
    "fela",
    "fela-tool",
    "cssinjs"
  ],
  "author": "Robin Weser",
  "license": "MIT",
  "peerDependencies": {
    "fela": "^11.2.0"
  },
  "dependencies": {
    "css-in-js-utils": "^3.0.0",
    "fast-loops": "^1.0.0",
    "fela": "^11.4.0",
    "fela-utils": "^11.4.0"
  },
  "gitHead": "4cc4686d9e16877251ac5c73e161076ec7c113a1"
}
