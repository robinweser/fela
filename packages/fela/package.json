{
  "name": "fela",
  "version": "11.4.0",
  "description": "State-Driven Styling in JavaScript",
  "typings": "index.d.ts",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "sideEffects": false,
  "scripts": {
    "build:umd": "webpack --config webpack.config.js"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.d.ts",
    "lib/**",
    "es/**",
    "umd/**"
  ],
  "repository": "https://github.com/robinweser/fela/",
  "keywords": [
    "fela",
    "dynamic styling",
    "stylesheet",
    "css",
    "styling",
    "functional styling",
    "functional css",
    "atomic css",
    "state-driven",
    "atomic design",
    "state-driven styling",
    "cssinjs"
  ],
  "author": "Robin Weser",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "^3.0.0",
    "csstype": "^2.5.5",
    "fast-loops": "^1.0.0",
    "fela-utils": "^11.4.0",
    "isobject": "^3.0.1"
  },
  "gitHead": "4cc4686d9e16877251ac5c73e161076ec7c113a1"
}
