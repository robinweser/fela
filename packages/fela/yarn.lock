# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


css-in-js-utils@2.0.0, css-in-js-utils@^2.0.0:
  version "2.0.0"
  resolved "https://registry.yarnpkg.com/css-in-js-utils/-/css-in-js-utils-2.0.0.tgz#5af1dd70f4b06b331f48d22a3d86e0786c0b9435"
  dependencies:
    hyphenate-style-name "^1.0.2"

fast-loops@^1.0.0:
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/fast-loops/-/fast-loops-1.0.0.tgz#2cdd7e0ff67343b2b5f5e627d855a50b4bed559a"

fela-tools@^5.1.2:
  version "5.1.2"
  resolved "https://registry.yarnpkg.com/fela-tools/-/fela-tools-5.1.2.tgz#c6760b42d72116911c2c0341b6d880cfc11c76a8"
  dependencies:
    css-in-js-utils "2.0.0"
    fast-loops "^1.0.0"
    fela "^6.1.3"
    fela-utils "^8.0.3"

fela-utils@^8.0.3:
  version "8.0.3"
  resolved "https://registry.yarnpkg.com/fela-utils/-/fela-utils-8.0.3.tgz#4a1728e7f5711e26f6f37da9c85a6b270c84c466"
  dependencies:
    css-in-js-utils "^2.0.0"
    fast-loops "^1.0.0"

fela@^6.1.3:
  version "6.1.3"
  resolved "https://registry.yarnpkg.com/fela/-/fela-6.1.3.tgz#c3bc8726dada63c23914bed9ca79d902e61e9a74"
  dependencies:
    css-in-js-utils "2.0.0"
    fast-loops "^1.0.0"
    fela-utils "^8.0.3"

hyphenate-style-name@^1.0.2:
  version "1.0.2"
  resolved "https://registry.yarnpkg.com/hyphenate-style-name/-/hyphenate-style-name-1.0.2.tgz#31160a36930adaf1fc04c6074f7eb41465d4ec4b"

string-hash@^1.1.3:
  version "1.1.3"
  resolved "https://registry.yarnpkg.com/string-hash/-/string-hash-1.1.3.tgz#e8aafc0ac1855b4666929ed7dd1275df5d6c811b"
