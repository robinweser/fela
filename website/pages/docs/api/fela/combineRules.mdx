# combineRules

A common pattern when styling elements is to have some base styles which get applied to every instance as well as some extended styles which get added based on some condition. Just concatenating classNames may lead to specificity problems as the last rendered rule always wins.<br />
To solve this issue, `combineRules` allows you to compose multiple rules into a single super selector.<br />

Combining rules actually is the best practice in terms of style composition as it prevents property specificity issues by default.

## Arguments

Accepts a list of [Rules](../../advanced/Rules.md).

## Returns

(_Function_) A super selector which composes all `rules` from left to right.

## Example

```javascript
import { combineRules } from 'fela'

const renderer = createRenderer()

const rule = props => ({
  fontSize: props.fontSize
  color: 'red'
})

const anotherRule = props => ({
  color: 'blue'
})

const superRule = combineRules(rule, anotherRule)

renderer.renderRule(rule, { fontSize: '12px '}) // => a b
renderer.renderRule(anotherRule) // => c
renderer.renderRule(superRule, { fontSize: '12px' }) // => a c
```

The following CSS would be rendered:

```css
.a {
  font-size: 12px;
}
.b {
  color: red;
}
.c {
  color: blue;
}
```
