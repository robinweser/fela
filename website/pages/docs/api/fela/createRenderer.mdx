import DocLayout from '../../../../components/DocLayout.js'

export default DocLayout

# createRenderer

Creates a Fela renderer which renders your selectors, keyframes, fonts and static styles.<br />
It caches all rendered styles to be able to reuse them on future rendering cycles.<br />

## Arguments

| Argument | Type      | Description                                                                                                                                                                                                                                                        |
| -------- | --------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| config   | _Object?_ | Optional renderer configuration. The most common use case is adding [plugins](../../advanced/Plugins.md) to process styles before they get cached.<br />_See [Advanced - Renderer Configuration](../../advanced/RendererConfiguration.md) for further information_ |

## Returns

([Renderer](Renderer.md)): A Renderer instance.

## Example

```javascript
import { createRenderer } from 'fela'

const rule = (props) => ({
  backgroundColor: 'red',
  fontSize: props.size,
  color: 'blue',
})

const renderer = createRenderer()

renderer.renderRule(rule, { size: '12px' }) // => a b c
```
