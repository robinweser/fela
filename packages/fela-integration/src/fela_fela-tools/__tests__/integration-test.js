import { createRenderer } from 'fela'
import { renderToString } from 'fela-tools'

describe('Fela with Fela Tools integration', () => {
  describe('Rendering rules', () => {
    it('should remove undefined values', () => {
      const rule = props => ({
        color: props.color,
        fontSize: '15px',
      })
      const renderer = createRenderer()

      const className = renderer.renderRule(rule)

      expect(className).toMatchSnapshot()
      expect(renderToString(renderer)).toMatchSnapshot()
    })

    it('should render pseudo classes', () => {
      const rule = () => ({
        color: 'red',
        ':hover': {
          color: 'blue',
        },
      })

      const renderer = createRenderer()
      renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
    })

    it('should prefix classNames', () => {
      const rule = () => ({
        color: 'red',
      })

      const renderer = createRenderer({
        selectorPrefix: 'fela_',
      })
      const className = renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
      expect(className).toMatchSnapshot()
    })
    it('should add specificity prefix', () => {
      const rule = () => ({
        color: 'red',
      })

      const renderer = createRenderer({
        specificityPrefix: '.parent ',
      })
      const className = renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
    })

    it('should render attribute selectors', () => {
      const rule = () => ({
        color: 'red',
        '[bool=true]': {
          color: 'blue',
        },
      })
      const renderer = createRenderer()

      renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
    })
    it('should render attribute selectors and add specificity prefix', () => {
      const rule = () => ({
        color: 'red',
        '[bool=true]': {
          color: 'blue',
        },
      })
      const renderer = createRenderer({ specificityPrefix: '.parent ' })

      renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
    })

    it('should render child selectors', () => {
      const rule = () => ({
        color: 'red',
        '>div': {
          color: 'blue',
        },
      })
      const renderer = createRenderer()

      renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
    })

    it('should render pseudo class selectors', () => {
      const rule = () => ({
        color: 'red',
        ':hover': {
          color: 'blue',
        },
      })
      const renderer = createRenderer()

      renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
    })

    it('should render any nested selector with the &-prefix', () => {
      const rule = () => ({
        color: 'red',
        '&~#foo': {
          color: 'blue',
        },
        '& .bar': {
          color: 'green',
        },
      })
      const renderer = createRenderer()

      renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
    })

    it('should render media queries', () => {
      const rule = () => ({
        color: 'red',
        '@media (min-height:300px)': {
          color: 'blue',
        },
      })

      const renderer = createRenderer()
      renderer.renderRule(rule)

      expect(renderToString(renderer)).toMatchSnapshot()
    })
  })

  describe('Rendering keyframes', () => {
    it('should render dynamic keyframe variations', () => {
      const keyframe = props => ({
        from: {
          color: props.color,
        },
        to: {
          color: 'blue',
        },
      })
      const renderer = createRenderer()

      const animationName = renderer.renderKeyframe(keyframe, {
        color: 'red',
      })

      expect(animationName).toMatchSnapshot()
      expect(renderToString(renderer)).toMatchSnapshot()
    })
  })

  describe('Rendering static styles', () => {
    it('should cache the style and return the rendered markup', () => {
      const renderer = createRenderer()

      const staticStyle = '*{color:red;margin:0}'
      renderer.renderStatic(staticStyle)

      expect(renderer.cache.hasOwnProperty(staticStyle)).toEqual(true)
      expect(renderToString(renderer)).toEqual(staticStyle)
    })

    it('should render a flat object of static selectors', () => {
      const renderer = createRenderer()

      const staticStyle = {
        margin: 0,
        fontSize: '12px',
      }

      renderer.renderStatic(staticStyle, 'html,body')
      expect(
        renderer.cache.hasOwnProperty('html,body{"margin":0,"fontSize":"12px"}')
      ).toEqual(true)
      expect(renderToString(renderer)).toMatchSnapshot()
    })

    it('should allow multiple static styles for a single selector', () => {
      const renderer = createRenderer()

      renderer.renderStatic(
        {
          margin: 0,
          fontSize: '12px',
        },
        'html,body'
      )
      renderer.renderStatic(
        {
          color: 'red',
        },
        'html,body'
      )

      expect(
        renderer.cache.hasOwnProperty('html,body{"margin":0,"fontSize":"12px"}')
      ).toEqual(true)
      expect(renderer.cache.hasOwnProperty('html,body{"color":"red"}')).toEqual(
        true
      )
      expect(renderToString(renderer)).toMatchSnapshot()
    })
  })
})
