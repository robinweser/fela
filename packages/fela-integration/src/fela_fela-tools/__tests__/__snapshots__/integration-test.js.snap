// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fela with Fela Tools integration Rendering keyframes should render dynamic keyframe variations 1`] = `"k1"`;

exports[`Fela with Fela Tools integration Rendering keyframes should render dynamic keyframe variations 2`] = `"@-webkit-keyframes k1{from{color:red}to{color:blue}}@-moz-keyframes k1{from{color:red}to{color:blue}}@keyframes k1{from{color:red}to{color:blue}}"`;

exports[`Fela with Fela Tools integration Rendering rules should add specificity prefix 1`] = `".parent .a{color:red}"`;

exports[`Fela with Fela Tools integration Rendering rules should prefix classNames 1`] = `".fela_a{color:red}"`;

exports[`Fela with Fela Tools integration Rendering rules should prefix classNames 2`] = `"fela_a"`;

exports[`Fela with Fela Tools integration Rendering rules should remove undefined values 1`] = `"a"`;

exports[`Fela with Fela Tools integration Rendering rules should remove undefined values 2`] = `".a{font-size:15px}"`;

exports[`Fela with Fela Tools integration Rendering rules should render any nested selector with the &-prefix 1`] = `".a{color:red}.b~#foo{color:blue}.c .bar{color:green}"`;

exports[`Fela with Fela Tools integration Rendering rules should render attribute selectors 1`] = `".a{color:red}.b[bool=true]{color:blue}"`;

exports[`Fela with Fela Tools integration Rendering rules should render attribute selectors and add specificity prefix 1`] = `".parent .a{color:red}.parent .b[bool=true]{color:blue}"`;

exports[`Fela with Fela Tools integration Rendering rules should render child selectors 1`] = `".a{color:red}.b>div{color:blue}"`;

exports[`Fela with Fela Tools integration Rendering rules should render media queries 1`] = `".a{color:red}@media (min-height:300px){.b{color:blue}}"`;

exports[`Fela with Fela Tools integration Rendering rules should render pseudo class selectors 1`] = `".a{color:red}.b:hover{color:blue}"`;

exports[`Fela with Fela Tools integration Rendering rules should render pseudo classes 1`] = `".a{color:red}.b:hover{color:blue}"`;

exports[`Fela with Fela Tools integration Rendering static styles should allow multiple static styles for a single selector 1`] = `"html,body{margin:0;font-size:12px}html,body{color:red}"`;

exports[`Fela with Fela Tools integration Rendering static styles should render a flat object of static selectors 1`] = `"html,body{margin:0;font-size:12px}"`;
